envoy/matcher/matcher.h

```diff
@@ -187,63 +196,67 @@ class InputMatcherFactory : public Config::TypedFactory {
 // Conceptually the data availability should start at being NotAvailable, transition to
 // MoreDataMightBeAvailable (optional, this doesn't make sense for all data) and finally
 // AllDataAvailable as the data becomes available.
-struct DataInputGetResult {
-  enum class DataAvailability {
-    // The data is not yet available.
-    NotAvailable,
-    // Some data is available, but more might arrive.
-    MoreDataMightBeAvailable,
-    // All the data is available.
-    AllDataAvailable
-  };
-
+template <class ResultDataType = std::string> struct DataInputGetResult {
   DataAvailability data_availability_;
   // The resulting data. This will be absl::nullopt if we don't have sufficient data available (as
   // per data_availability_) or because no value was extracted. For example, consider a DataInput
   // which attempts to look a key up in the map: if we don't have access to the map yet, we return
   // absl::nullopt with NotAvailable. If we have the entire map, but the key doesn't exist in the
   // map, we return absl::nullopt with AllDataAvailable.
-  absl::optional<std::string> data_;
+  absl::optional<ResultDataType> data_;
 
   // For pretty printing.
   friend std::ostream& operator<<(std::ostream& out, const DataInputGetResult& result) {
-    out << "data input: " << (result.data_ ? result.data_.value() : "n/a");
+    if (result.data_.has_value()) {
+      out << "data input: " << result.data_.value();
+    } else {
+      out << "data input: n/a";
+    }
+
     switch (result.data_availability_) {
-    case DataInputGetResult::DataAvailability::NotAvailable:
+    case DataAvailability::NotAvailable:
       out << " (not available)";
       break;
-    case DataInputGetResult::DataAvailability::MoreDataMightBeAvailable:
+    case DataAvailability::MoreDataMightBeAvailable:
       out << " (more data available)";
       break;
-    case DataInputGetResult::DataAvailability::AllDataAvailable:;
+    case DataAvailability::AllDataAvailable:;
     }
     return out;
   }
 };
 
+// Template deduction guide for DataInputGetResult.
+template <class ResultDataType>
+DataInputGetResult(DataAvailability data_availability, ResultDataType data)
+    -> DataInputGetResult<std::string>;
```

Comment:
I'm not sure I understand this guide - it sounds like it's saying that any constructed `DataInputGetResult` is actually `DataInputGetResult<string>`?

Reply:
This guide to help compiler in certain cases to deduce the template arguments without specifying `std::string` , For instance, we don't need to have `DataInputGetResult<std::string>` in the createMatchers field_matcher_test.cc.

If we want to use other data type, we just need to specify `ResultDataType`. The use case for such will be added in my following PRs when add the support for CEL matcher with this matching API.

A quick draft example in same field_matcher_test.cc for illustrative purpose, in addition to existing string `createMatchers`, we can have 
```
std::vector<FieldMatcherPtr<TestIntData>>
  createIntMatchers(std::vector<std::pair<bool, DataAvailability>> values) {
    std::vector<FieldMatcherPtr<TestIntData>> matchers;
    matchers.reserve(values.size());
    for (const auto& v : values) {
      matchers.emplace_back(std::make_unique<SingleFieldMatcher<TestIntData, int>>(
          std::make_unique<TestIntInput>(DataInputGetResult<int>{v.second, absl::nullopt}),
          std::make_unique<BoolMatcher>(v.first)));
    }
    return matchers;
  }
```

Reply:
Yeah, I think I see it as a default compiler hint - meaning it'll try to deduce string as soon as possible, and that means any other type value but the string requires a copy of code - `createIntMatchers` instead of generic `createMatchers`. I don't have a clear grasp of how much code needs to be explicitly instantiated since all inputs are strings -- hence, my concern.

Reply:
`createIntMatchers` here is the illustrative example (and in test code). I think If it is set to be int type, there will be no need of the impl for the other type.

From my prototype for CEL with all the changes requires, there will not have copy of code in the non-test code path as they are all handled in place with templatized  argument and we just need to specify the template argument when we use it. I will share more details in the doc i mentioned below.

Reply:
Ack, thanks for confirming.